ipaplatform/redhat/authconfig.py

```diff
@@ -129,7 +129,14 @@ def configure(self, sssd, mkhomedir, statestore, sudo=True):
     def unconfigure(
         self, fstore, statestore, was_sssd_installed, was_sssd_configured
     ):
-        if not statestore.has_state('authselect') and was_sssd_installed:
+        # If the installation failed before doing the authselect part
+        # nothing to do here
+        complete = statestore.get_state('installation', 'complete')
+        if complete is not None and not complete and \
+           not statestore.has_state('authselect'):
+            return
```

Comment:
This code is a bit blocked by the is_ipa_client_configured() call which is very strict: it is either configured completely or not at all which is likely another bug.
But if we get past that and complete == False and authselect is not in the state file then this will spew the same WARNING that a cited patch attempted to remove.
This is not possible now but at some point we may fix the is_ipa_client_configured() to be a bit more flexible to provide a semi-installed state (likely as None). For example, if you run ipa-client-install and ^C before it's done then configured=False and there may be other state depending on what you did, but you can't uninstall anything because it isn't configured (complete != True).

Reply:
I'm not sure I get your point about being blocker by is_ipa_client_configured. I ran the test `test_replica_promotion.py::TestWrongClientDomain::test_client_rollback` with my fix, and it didn't show any warning (see the [report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/c8febfd2-cd3f-11ec-a75e-fa163e4fd377/report.html)):
```
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:513 RUN ['ipa-client-install', '-U', '--server', 'master.ipa.test', '--domain', 'ipa.test', '-w', 'foo']
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Client hostname: replica0.ipa.test
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Realm: IPA.TEST
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 DNS Domain: ipa.test
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 IPA Server: master.ipa.test
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 BaseDN: dc=ipa,dc=test
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Synchronizing time
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 No SRV records of NTP servers found and no NTP server or pool address was provided.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Attempting to sync time with chronyc.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Process chronyc waitsync failed to sync time!
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Unable to sync time with chrony server, assuming the time is in sync. Please check that 123 UDP port is opened, and any time server is on network.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Downloading the CA certificate via HTTP, this is INSECURE
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Successfully retrieved CA cert
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557     Subject:     CN=Certificate Authority,O=IPA.TEST
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557     Issuer:      CN=Certificate Authority,O=IPA.TEST
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557     Valid From:  2022-05-06 13:29:24
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557     Valid Until: 2042-05-06 13:29:24
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Joining realm failed: SASL Bind failed
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557     Invalid credentials
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Installation failed. Rolling back changes.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Disabling client Kerberos and LDAP configurations
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 nscd daemon is not installed, skip configuration
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 nslcd daemon is not installed, skip configuration
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Client uninstall complete.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 The ipa-client-install command failed. See /var/log/ipaclient-install.log for more information
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 This program will set up IPA client.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Version 4.10.0.dev
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:557 Using default chrony configuration.
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica0.cmd40:transport.py:217 Exit code: 1
```

The method `unconfigure()` can be called from 2 different paths:
- from the uninstaller (and in this case, yes, is_ipa_client_configured can stop the execution early, even before this point is reached)
- from the installer if one of the steps failed (there is a try: _install() except: uninstall())
The diff here targets this 2nd case.

Reply:
Sure but if _install() fails prior to authselect state being set then the new condition will fail and fall through to the message we want to suppress right?

Reply:
If _install() fails prior to authselect state being set, then we go to line 137 where there is `return`, no message is printed and nothing is reverted as there is nothing to do. Am I missing something?

Reply:
Ok, I'm convinced. I'm paranoid because this code has proved to be fragile in the past.
Can you update the commit message to not put so much blame on https://github.com/freeipa/freeipa/commit/e59ee6099f065c2155e3be13025459108daf7900 ? This code pre-dated the complete state by 2 years so there was no way to know whether this was removing a failed or successful installation.

Reply:
Sure, I updated the message.